### ISSUE TITLE
linux vagrant boxes hang during testing

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#5146]
It seems there is an issue with not unmounting stuff in the linux machines and thus running out of loop devices:

```
vagrant@ubuntu-bionic:/vagrant/borg/borg$ mount
...
borgfs on /tmp/tmpqxvs1fmw type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmp9a_cpv4r type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmpahx7c7mb type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmp5jm22hqr type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmpyofzz0r8 type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmp9rt4haej type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmpu43lfcz4 type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
borgfs on /tmp/tmp3hk9yo7s type fuse (ro,nosuid,nodev,relatime,user_id=1000,group_id=1000,default_permissions)
```
```
vagrant@ubuntu-bionic:/vagrant/borg/borg$ ls -l /dev/loop*
brw-rw---- 1 root disk  7,   0 Apr 20 11:49 /dev/loop0
brw-rw---- 1 root disk  7,   1 Apr 20 11:49 /dev/loop1
brw-rw---- 1 root disk  7,   2 Apr 20 11:49 /dev/loop2
brw-rw---- 1 root disk  7,   3 Apr 20 11:49 /dev/loop3
brw-rw---- 1 root disk  7,   4 Apr 20 11:49 /dev/loop4
brw-rw---- 1 root disk  7,   5 Apr 20 11:49 /dev/loop5
brw-rw---- 1 root disk  7,   6 Apr 20 11:49 /dev/loop6
brw-rw---- 1 root disk  7,   7 Apr 20 11:49 /dev/loop7
crw-rw---- 1 root disk 10, 237 Apr 20 11:49 /dev/loop-control
```

**Comment by ThomasWaldmann** [comment]
The state shown above is while the tests just hang doing nothing.

If I manually umount some borgfs filesystems in parallel, that solves the hang and the tests continue.

**Comment by ThomasWaldmann** [comment]
@finefoot ^

**Comment by ThomasWaldmann** [comment]
Maybe related to fakeroot usage.

Without fakeroot, it succeeds:
```
pytest -k test_readonly_check -v  # no problem, tests succeed
```

With fakeroot, it fails:
```
fakeroot -u pytest -k test_readonly_check -v  # fails, see below
```

```
self = <borg.testsuite.archiver.ArchiverTestCase testMethod=test_readonly_check>

    def test_readonly_check(self):
        self.cmd('init', '--encryption=repokey', self.repository_location)
        self.create_src_archive('test')
        with self.read_only(self.repository_path):
            # verify that command normally doesn't work with read-only repo
            if self.FORK_DEFAULT:
                self.cmd('check', '--verify-data', self.repository_location, exit_code=EXIT_ERROR)
            else:
                with pytest.raises((LockFailed, RemoteRepository.RPCError)) as excinfo:
>                   self.cmd('check', '--verify-data', self.repository_location)
E                   Failed: DID NOT RAISE (<class 'borg.locking.LockFailed'>, <class 'borg.remote.RemoteRepository.RPCError'>)

src/borg/testsuite/archiver.py:1541: Failed
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
chattr: Operation not permitted while setting flags on /tmp/tmpf1mo6i57/repository
```

**Comment by ThomasWaldmann** [comment]
Guess chmod does not work because we are fake root and chattr does not work, because we are only fake root, not real root?

**Comment by ThomasWaldmann** [comment]
That would avoid the issue. @finefoot do you have a better idea?
```
diff --git a/src/borg/testsuite/archiver.py b/src/borg/testsuite/archiver.py
index 3bf65216..e56a27f3 100644
--- a/src/borg/testsuite/archiver.py
+++ b/src/borg/testsuite/archiver.py
@@ -1529,6 +1529,7 @@ def test_corrupted_repository(self):
         output = self.cmd('check', '--info', self.repository_location, exit_code=1)
         self.assert_in('Starting repository check', output)  # --info given for root logger
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_check(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('test')
@@ -1544,6 +1545,7 @@ def test_readonly_check(self):
             # verify that command works with read-only repo when using --bypass-lock
             self.cmd('check', '--verify-data', self.repository_location, '--bypass-lock')
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_diff(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('a')
@@ -1560,6 +1562,7 @@ def test_readonly_diff(self):
             # verify that command works with read-only repo when using --bypass-lock
             self.cmd('diff', '%s::a' % self.repository_location, 'b', '--bypass-lock')
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_export_tar(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('test')
@@ -1575,6 +1578,7 @@ def test_readonly_export_tar(self):
             # verify that command works with read-only repo when using --bypass-lock
             self.cmd('export-tar', '%s::test' % self.repository_location, 'test.tar', '--bypass-lock')
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_extract(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('test')
@@ -1590,6 +1594,7 @@ def test_readonly_extract(self):
             # verify that command works with read-only repo when using --bypass-lock
             self.cmd('extract', '%s::test' % self.repository_location, '--bypass-lock')
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_info(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('test')
@@ -1605,6 +1610,7 @@ def test_readonly_info(self):
             # verify that command works with read-only repo when using --bypass-lock
             self.cmd('info', self.repository_location, '--bypass-lock')
 
+    @pytest.mark.skipif(fakeroot_detected(), reason='readonly tests are not compatible with fakeroot')
     def test_readonly_list(self):
         self.cmd('init', '--encryption=repokey', self.repository_location)
         self.create_src_archive('test')
```

**Comment by finefoot** [comment]
My experience with fakeroot is limited. But now that I think about it, it might be possible that fakeroot automatically avoids withdrawing permissions because (real) root cannot do that.

**Comment by finefoot** [comment]
I'll look into it and report back.

**Comment by ThomasWaldmann** [comment]
See #5147 - I'll merge that for now.

If there's a better solution later, we can merge it later.

**Comment by finefoot** [comment]
I found something. Writing a fix right now.

**Comment by finefoot** [comment]
@ThomasWaldmann See linked PR

**Comment by ThomasWaldmann** [comment]
Ah, that was quick. I'll try it in my vagrant VMs...

**Comment by ThomasWaldmann** [comment]
Fixed by #5148.
